### PR TITLE
[Bugfix] Fix MTP acceptance regression on plan stream by moving int64 cast before plan stream context

### LINKED ISSUE #28405: [Bug] [NPU] When enable plan stream, the acceptance rate of MTP decreases
### Checklist

- [x] I searched related issues but found no solution.
- [x] The bug persists in the latest version.
- [ ] Issues without environment info and a minimal reproducible demo are hard to resolve and may receive no feedback.
- [ ] If this is not a bug report but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

### Describe the bug

When enabling Plan Stream via SGLANG_ENABLE_OVERLAP_PLAN_STREAM=1, the acceptance rate of MTP decreases, while model accuracy remains unaffected.

```
[2026-06-16 16:31:47 TP0] Decode batch, #running-req: 63, #full token: 55552, full token usage: 0.05, mamba num: 63, mamba usage: 0.98, accept len: 1.32, accept rate: 0.11, npu graph: True, gen throughput (token/s): 492.98, #queue-req: 0

┌─────────────────┬───────────┬──────────┬──────────┬───────┬─────────┬─────────┐
│ Model           │ Dataset   │ Metric   │ Subset   │   Num │   Score │ Cat.0   │
├─────────────────┼───────────┼──────────┼──────────┼───────┼─────────┼─────────┤
│ Qwen3.6-35B-A3B │ gsm8k     │ mean_acc │ main     │   200 │   0.965 │ default │
└─────────────────┴───────────┴──────────┴──────────┴───────┴─────────┴─────────┘
```

### Reproduction

```
export ASCEND_USE_FIA=1
export SGLANG_ENABLE_SPEC_V2=1
export SGLANG_ENABLE_OVERLAP_PLAN_STREAM=1

MODEL_PATH=/home/weights/Qwen3.6-35B-A3B
sglang serve \
        --model-path ${MODEL_PATH} \
        --attention-backend ascend \
        --device npu \
        --tp-size 2 --nnodes 1 --node-rank 0 \
        --chunked-prefill-size -1 \
        --trust-remote-code \
        --host 127.0.0.1 --max-running-requests 64 \
        --mem-fraction-static 0.8 \
        --port 9903 \
        --mm-attention-backend ascend_attn \
        --dtype bfloat16 --mamba-ssm-dtype bfloat16 \
        --base-gpu-id 12 \
        --enable-multimodal \
        --disable-radix-cache \
        --speculative-algorithm NEXTN --speculative-num-steps 3 --speculative-eagle-topk 1 --speculative-num-draft-tokens 4 \
        --speculative-draft-model-quantization unquant \
        --cuda-graph-bs 8 12 16 20 24 32 40 48 56 64 \
```

### Environment

Python: 3.11.15 (main, May 13 2026, 09:39:24) [GCC 11.4.0]
NPU available: True
NPU 0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15: Ascend910_9362
CANN_HOME: /usr/local/Ascend/cann-9.0.0
CANN: Not Available
BiSheng: 2026-04-25T15:46:25+08:00 clang version 15.0.5 (clang-5c68a1cb1231 flang-5c68a1cb1231)
Ascend Driver Version: 25.5.1
PyTorch: 2.10.0+cpu
sglang: 0.5.12.post2.dev755+g5ae8d286d

### PR DESCRIPTION
**Pull request by iridiumine**
## Motivation
Closes https://github.com/sgl-project/sglang/issues/28405, https://github.com/sgl-project/sglang/issues/28406.
PR #28043 fixed a DeepSeek-V4 DP dtype mismatch hang by normalizing predict to torch.int64 inside prepare_for_draft_extend . However, when the overlap plan stream is enabled, this cast executes on the plan stream while predict is produced on the main stream, creating a cross-stream data race that silently corrupts draft token ids and drops MTP acceptance to nearly zero.

## Modifications
- eagle_worker_v2.py : Move the predict.to(torch.int64) cast from inside prepare_for_draft_extend to _draft_extend_for_decode , before entering the plan stream context. The cast now runs on the main stream with correct synchronization.
- base_spec_worker.py : Update the comment in prepare_for_draft_extend to explain why dtype normalization must NOT be done inside this function, and that the caller should cast before entering the plan stream.

## Accuracy Tests
No impact on model output — the int64 cast preserves all token id values. The fix only changes where the cast executes (main stream vs. plan stream).

```
export ASCEND_USE_FIA=1
export SGLANG_ENABLE_SPEC_V2=1
export SGLANG_ENABLE_OVERLAP_PLAN_STREAM=1

MODEL_PATH=/home/weights/Qwen3.6-35B-A3B
sglang serve \
        --model-path ${MODEL_PATH} \
        --attention-backend ascend \
        --device npu \
        --tp-size 2 --nnodes 1 --node-rank 0 \
        --chunked-prefill-size -1 \
        --trust-remote-code \
        --host 127.0.0.1 --max-running-requests 64 \
        --mem-fraction-static 0.8 \
        --port 9903 \
        --mm-attention-backend ascend_attn \
        --dtype bfloat16 --mamba-ssm-dtype bfloat16 \
        --base-gpu-id 12 \
        --enable-multimodal \
        --disable-radix-cache \
        --speculative-algorithm NEXTN --speculative-num-steps 3 --speculative-eagle-topk 1 --speculative-num-draft-tokens 4 \
        --speculative-draft-model-quantization unquant \
        --cuda-graph-bs 8 12 16 20 24 32 40 48 56 64 \
```

```
┌─────────────────┬───────────┬──────────┬──────────┬───────┬─────────┬─────────┐
│ Model           │ Dataset   │ Metric   │ Subset   │   Num │   Score │ Cat.0   │
├─────────────────┼───────────┼──────────┼──────────┼───────┼─────────┼─────────┤
│ Qwen3.6-35B-A3B │ gsm8k     │ mean_acc │ main     │   200 │    0.96 │ default │
└─────────────────┴───────────┴──────────┴──────────┴───────┴─────────┴─────────┘
```

## Speed Tests and Profiling
No impact on inference speed. The int64 cast is a lightweight element-wise operation on a small token-id tensor, not on hidden states, logits, or KV cache. The cast already existed in PR #28043; this PR only moves its execution point without adding extra overhead.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #27836569566](https://github.com/sgl-project/sglang/actions/runs/27836569566)
Latest PR Test (Extra): :x: [Run #27836569483](https://github.com/sgl-project/sglang/actions/runs/27836569483)

### CODE REVIEW

**Comment by AndyLi429** [comment]
LGTM

File: python/sglang/srt/speculative/eagle_worker_v2.py [hunk 1]
```diff
@@ -748,12 +748,16 @@ def _draft_extend_for_decode(
...
+        # synchronization issues with .to() inside the plan stream context.
+        next_token_ids = batch_result.next_token_ids.to(torch.int64)
+
         # Prepare for draft extend in a separate stream
         with self.plan_stream_ctx:
             forward_batch = self.prepare_for_draft_extend(
                 draft_extend_input,
                 batch,
-                batch_result.next_token_ids,
+                next_token_ids,
```
Comment (AndyLi429) [inline python/sglang/srt/speculative/eagle_worker_v2.py:760]: only this function use prepare_for_draft_extend?

Comment (iridiumine) [reply python/sglang/srt/speculative/eagle_worker_v2.py:760]: Yes, only these two call sites. multi_layer_eagle_worker_v2.py passes batch_result.next_token_ids directly without .to(int64) , so no copy is created inside the plan stream and no data race. No fix needed there.

**Comment by DaZhUUU** [comment]
Thanks for catching this, and sorry I missed it. 

#28043 was focused on the DP collective dtype mismatch/hang, and I only validated request success plus NCCL dtype behavior, not MTP acceptance.

**Review by iforgetmyname — APPROVED** [review]
